components/eamxx/src/physics/iop_forcing/eamxx_iop_forcing_process_interface.cpp

```diff
@@ -0,0 +1,546 @@
+
+    // At the top and bottom of the model, set the end points for
+    // delta_*_k and delta_*_km1 to be the first and last entries
+    // of delta_*, respectively.
+    if (any_at_top) {
+      delta_u_k.set(at_top, s_delta_u(0));
+      delta_v_k.set(at_top, s_delta_v(0));
+      delta_T_k.set(at_top, s_delta_T(0));
+    }
+    if (any_at_bot) {
+      delta_u_km1.set(at_bot, s_delta_u(nlevs-2));
+      delta_v_km1.set(at_bot, s_delta_v(nlevs-2));
+      delta_T_km1.set(at_bot, s_delta_T(nlevs-2));
+    }
+
+    // Get omega_int(k+1) pack. The range pack should not
+    // contain index > nlevs-1 (since omega_int is size nlevs+1).
+    auto range_pack_for_p1_shift = range_pack;
+    range_pack_for_p1_shift.set(range_pack>nlevs-1, nlevs-1);
+    Pack omega_int_k, omega_int_kp1;
+    ekat::index_and_shift<1>(s_omega_int, range_pack, omega_int_k, omega_int_kp1);
+
+    const auto fac = (dt/2)/ref_p_del(k);
+
+    // Update u
+    u(k).update(not_at_bot, fac*omega_int_kp1*delta_u_k, -1, 1);
+    u(k).update(not_at_top, fac*omega_int_k*delta_u_km1, -1, 1);
+
+    // Update v
+    v(k).update(not_at_bot, fac*omega_int_kp1*delta_v_k, -1, 1);
+    v(k).update(not_at_top, fac*omega_int_k*delta_v_km1, -1, 1);
+
+    // Before updating T, first scale using thermal
+    // expansion term due to LS vertical advection
+    T(k) *= 1 + (dt*Rair/Cpair)*omega(k)/ref_p_mid(k);
+
+    // Update T
+    T(k).update(not_at_bot, fac*omega_int_kp1*delta_T_k, -1, 1);
+    T(k).update(not_at_top, fac*omega_int_k*delta_T_km1, -1, 1);
+
+    // Update Q
+    Pack delta_tracer_k, delta_tracer_km1;
+    for (int iq=0; iq<n_q_tracers; ++iq) {
+      auto s_delta_tracer = Kokkos::subview(s_delta_Q, iq, Kokkos::ALL());
+      ekat::index_and_shift<-1>(s_delta_tracer, range_pack_for_m1_shift, delta_tracer_k, delta_tracer_km1);
+      if (any_at_top) delta_tracer_k.set(at_top, s_delta_tracer(0));
+      if (any_at_bot) delta_tracer_km1.set(at_bot, s_delta_tracer(nlevs-2));
+
+      Q(iq, k).update(not_at_bot, fac*omega_int_kp1*delta_tracer_k, -1, 1);
+      Q(iq, k).update(not_at_top, fac*omega_int_k*delta_tracer_km1, -1, 1);
+    }
+  });
+
+  // Release WS views
+  workspace.release_macro_block(delta_Q_slot, n_q_tracers);
+  workspace.release_many_contiguous<4>({&omega_int,  &delta_u,  &delta_v,  &delta_T});
+}
+// =========================================================================================
+KOKKOS_FUNCTION
+void IOPForcing::
+advance_iop_forcing(const MemberType& team,
+                         const int nlevs,
+                         const Real dt,
+                         const view_1d<const Pack>& divT,
+                         const view_1d<const Pack>& divq,
+                         const view_1d<Pack>& T,
+                         const view_1d<Pack>& qv)
+{
+  const auto nlev_packs = ekat::npack<Pack>(nlevs);
+  Kokkos::parallel_for(Kokkos::TeamVectorRange(team, nlev_packs), [&] (const int k) {
+    T(k).update(divT(k), dt, 1.0);
```

Comment:
I remember there was an issue with state update, which had an extra NUM_MPI_TASKS factor somewhere. I assume that has been fixed?

Reply:
Yeah, this was an issue with the domain means below, where I was not using the global number of columns previously.